### PR TITLE
Add convertToNamedParameters support for Scala 3

### PR DESCRIPTION
**Pull request by jkciesluk**
Closes https://github.com/scalameta/metals/issues/4059
Add Scala 3 support for https://github.com/scalameta/metals/pull/3971

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/codeactions/ConvertToNamedArguments.scala [hunk 1]
```diff
@@ -44,47 +41,40 @@ class ConvertToNamedArguments(trees: Trees, buildTargets: BuildTargets)
   ): Future[Seq[l.CodeAction]] = {
 
     val path = params.getTextDocument().getUri().toAbsolutePath
-    val isScala3 = buildTargets
-      .scalaVersion(path)
-      .map(ScalaVersions.isScala3Version)
-      .getOrElse(false)
-    if (isScala3) Future.successful(Nil)
-    else {
```
Comment (tanishiking) [inline metals/src/main/scala/scala/meta/internal/metals/codeactions/ConvertToNamedArguments.scala:52]: :+1:

**Review by tanishiking — CHANGES_REQUESTED** [review]
Thank you for working on this @jkciesluk !

This implementation converts all Apply to named arguments instead of selected Apply, we have to convert only the selected one (in the following case, we should convert only the arguments of `foo` method).

named-convert

---

Also, we would need to copy some test suite from https://github.com/scalameta/metals/blob/ec89eac187c5c9e1358a8229ebac85b64a71a8a9/tests/unit/src/test/scala/tests/codeactions/ConvertToNamedArgumentsLspSuite.scala to https://github.com/scalameta/metals/blob/ec89eac187c5c9e1358a8229ebac85b64a71a8a9/tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala (I'm wondering we can cross run those LSP test 🤔  but not this time)

File: mtags/src/main/scala-3/scala/meta/internal/pc/ConvertToNamedArgumentsProvider.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                case ((arg, index), param) if argIndices.contains(index) => {
+                  val position = arg.sourcePos.toLSP
+                  position.setEnd(position.getStart())
+                  new l.TextEdit(position, s"${param.name.show} = ")
+                }
+              }
+              .foreach(result.addOne(_))
+          case t =>
+            traverseChildren(t)
+    traverser.traverse(tree)(using newctx)
```
Comment (tanishiking) [inline mtags/src/main/scala-3/scala/meta/internal/pc/ConvertToNamedArgumentsProvider.scala:48]: Instead of traversing the whole tree, it would be useful to check the enclosing tree (of the given position (via param)) using `Interactive.pathTo` or `Interactive.enclosingTree`

Comment (jkciesluk) [reply mtags/src/main/scala-3/scala/meta/internal/pc/ConvertToNamedArgumentsProvider.scala]: Hi, thank you for help! 
Using `pathTo` works. I've also copied tests to https://github.com/scalameta/metals/blob/ec89eac187c5c9e1358a8229ebac85b64a71a8a9/tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala and extended them by testing multiple Apply.

**Review by tanishiking — APPROVED** [review]
LGTM!

**Comment by tanishiking** [comment]
You may have to rebase on `main` and run scalafmt for https://github.com/scalameta/metals/pull/4050

File: tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala [hunk 3]
```diff
@@ -328,6 +338,264 @@ class Scala3CodeActionLspSuite
     expectNoDiagnostics = false,
   )
 
+  check(
```
Comment (tgodzik) [inline tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala:341]: We can really just add one test (first one looks good), since we only want to test the branch when we have Scala 3 dialect.  All the other cases are already handled in the ConvertToNamedArgumentsLspSuite.

Anything specific to Scala 3 we can test in cross suite.

The reason is for this comment is that these tests each can last up to a couple of seconds and it's harsh on out CI.

File: tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala [hunk 4]
```diff
@@ -12,12 +13,16 @@ import scala.meta.internal.mtags.MtagsEnrichments.XtensionAbsolutePath
...
 import munit.Location
 import munit.TestOptions
+import org.eclipse.lsp4j.CodeAction
 import tests.codeactions.BaseCodeActionLspSuite
 
 class Scala3CodeActionLspSuite
     extends BaseCodeActionLspSuite("cross-code-actions") {
 
   override protected val scalaVersion: String = BuildInfo.scala3
+  val filterAction: CodeAction => Boolean = { act: CodeAction =>
```
Comment (tgodzik) [inline tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala:23]: Why is this needed? I see that we anyway `ConvertToNamedArguments` action showing up below?

**Review by tgodzik — CHANGES_REQUESTED** [review]
Looks good! Just very minor remarks about the tests

File: tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala (hunk 3 above)
Comment (tanishiking) [reply tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala:338]: Ah, right.

> All the other cases are already handled in the ConvertToNamedArgumentsLspSuite.
`ConvertToNamedArgumentsLspSuite` test against Scala 2.13, it doesn't verify anything against Scala 3.

Actually, those tests that verify the behavior (rather than the first one) should be in PC tests instead of `xxxLSPSuite`.
Sorry, we should did this in https://github.com/scalameta/metals/pull/3971

For example, Implementing all abstract member code action 
 have two (three?) test suite 
- cross test in pc (presentation compiler) 
  - https://github.com/scalameta/metals/blob/53249aa4047ee304275a818d3995a4b56a0b42b2/tests/cross/src/test/scala/tests/pc/AutoImplementAbstractMembersSuite.scala#L14
  - These tests verify the behavior of code action, but it doesn't check the code action is available or not. Those test anyway invoke the code action even if the user won't see it's available.
  - These tests run against all Scala versions
  - You can run those tests with the specific version by `++3.1.3 cross/testOnly tests.pc.AutoImplementAbstractMemberSuite`.
- ImplementAbstractMemberLspSuite
  - https://github.com/scalameta/metals/blob/53249aa4047ee304275a818d3995a4b56a0b42b2/tests/unit/src/test/scala/tests/codeactions/ImplementAbstractMembersLspSuite.scala#L5-L6
  - These test verify the code actions are available for users when we have a certain error (missing implementation error)
  - This test runs against only Scala 2.13
- Scala3CodeActionLspSuite
  - https://github.com/scalameta/metals/blob/53249aa4047ee304275a818d3995a4b56a0b42b2/tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala#L332-L358
  - This test check the code action is available in Scala3. We test this because if the Scala3 compiler has different error message than Scala2, we might need to change the regex to capture the error message for missing implementation error.
  - This test runs against only Scala3.

---

So, we should migrate most of `ConvertToNamedArgumentsLspSuite` into pc suite (that we haven't yet implemented). but it can be done in another PR :)

**Review by tgodzik — APPROVED** [review]
LGTM!

**Review by tgodzik — APPROVED** [review]
LGTM!

File: tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala (hunk 3 above)
Comment (tgodzik) [reply tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala:339]: Scala3CodeActionLspSuite will run the entire server, while cross tests will only invoke the compiler, which should be much faster.

Comment (tgodzik) [reply tests/slow/src/test/scala/tests/feature/Scala3CodeActionLspSuite.scala:339]: It's probably fine to test if the workflow works for Scala 3, which confirms one new path for testing, while we can test anything specific to Scala 3 compiler in the cross tests.
